README
CliMax version 1.0
=======================

NAME:

Maximal Clique Computation Tool


DESCRIPTION:


Given an adjacency list the program computes all maximal cliques in a 
given graph (set of graphs)


SYNOPSIS:
    
      
    #include<vector>
    
    #include<Graph.hpp>
    #include<Clique.hpp>
    
    
/* Constructor */
 
/*NOTE:  parent -> vector<int|long|double>, child -> vector<int|long|double> */  

    CliMax<int|long|double> clique(parent, child, 3);  
  



/* Compute Maximal Cliques */  
      clique.ComputeMaxCliques();
      



/* Get Maximal cliques */
      vector<vector<int|long|double>> mc = clique.GetMaxCliques();
      /*  mc contains vector of vertices of a given maximal clique */







CONSTRUCTION AND RUNTIME COMPLEXITY:


     Worst case for n = |V| (number of vertices):
    
     O(3^(n/3))


ACKNOWLEDGMENTS:

Tomita, E.,Tanaka, A. and  Takahashi, H. The wrst-case time complexity 
for generating all maximal cliques and computational experiments. (2006) 
Theor. Comput. Sci. 363:28-42



AUTHOR:


Robert Bakaric <rbakaric@irb.hr>, <bakaric@evolbio.mpg.de>

COPYRIGHT AND LICENSE:

 * Copyright 2015 Robert Bakaric
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.



ChangeLog
2015-04-21  Robert Bakaric  <rbakaric@irb.hr>
	* Created! 


src/include/Graph.hpp
/*
 * GraphMan.hpp
 * 
 * Copyright 2015 Robert Bakaric <rbakaric@irb.hr>
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 * 
 * 
 */

#include <vector>
#include <unordered_map>

using namespace std;


/* Graph */

template <typename Tint>
class Graph{

  protected:
/* Data Container */
  unordered_map<Tint, vector<Tint>> GraphMap;
  unordered_map<Tint,Tint>          GraphMapRev;
  unordered_map<Tint,Tint>          Traversed;
  vector<Tint>                      TraversalPath;
  vector<Tint>                      Distance;
  Tint Max;

/* Functions */ 
  void MakeGraph(const vector<Tint>& parent, const vector<Tint>& child);
  
  public:

/* Constructor */
   Graph(const vector<Tint>& parent, const vector<Tint>& child);
/* Destructor */
  ~Graph();
/* Explicite Constructor */
   void make(const vector<Tint>& parent, const vector<Tint>& child);
/* Explicite destructor */
   void destroy();
  
/* Functions */
   void AddVertex(const Tint vertex, const Tint location); /* Note. to avoid cycles use CreateVertex() for vertex variable */
   void RemoveVertex(const Tint vertex);
   void JoinVertex(const Tint vertexA,const Tint vertexB);
   void RelocateVertex(const Tint vertex,const Tint parentvertex);
   void EraseVertex(const Tint vertex);
   void DepthFirstTrav(const Tint vertex, Tint depth, Tint Flag);
   /* BreadthFirstTrav() : NOT FINISHED YET */
   Tint CreateVertex();
   vector<Tint>& GetDepthVec();
   vector<Tint>& GetVerticesVec();
   void Clean();
   void DmpGraph(vector<Tint>& parent, vector<Tint>& child);
};



/* Constructors */

template <typename Tint>
Graph<Tint>::Graph(const vector<Tint>& parent, const vector<Tint>& child){
  Max = 0;
  MakeGraph(parent, child);
}

template <typename Tint>
void Graph<Tint>::make(const vector<Tint>& parent, const vector<Tint>& child){
  Max = 0;
  MakeGraph(parent, child);
}


template <typename Tint>
void Graph<Tint>::MakeGraph(const vector<Tint>& parent, const vector<Tint>& child){
  for (Tint i = 0; i< parent.size(); i++){
    
    if (child[i] > Max) Max = child[i];
    if (parent[i] > Max) Max = parent[i];
    
    GraphMap[parent[i]].push_back(child[i]);
    GraphMapRev[child[i]] = parent[i];
  }
}


/* Destructors */

template <typename Tint>
Graph<Tint>::~Graph(){
  GraphMap.clear();
  GraphMapRev.clear();
  Traversed.clear();
  vector<Tint>().swap(TraversalPath);
  vector<Tint>().swap(Distance);

}

template <typename Tint>
void Graph<Tint>::destroy(){
  GraphMap.clear();
  GraphMapRev.clear();
  Traversed.clear();
  vector<Tint>().swap(TraversalPath);
  vector<Tint>().swap(Distance);

}



/* Functions */

template <typename Tint>
Tint Graph<Tint>::CreateVertex(){
   return ++Max;
}


template <typename Tint>
void Graph<Tint>::AddVertex(const Tint vertex, const Tint location){
   GraphMap[location].push_back(vertex);
   GraphMapRev[vertex] = location;
}


template <typename Tint>
void Graph<Tint>::RemoveVertex(const Tint vertex){
   Tint parent = GraphMapRev[vertex];
   GraphMap[parent].erase( remove( begin(GraphMap[parent]), end(GraphMap[parent]), vertex ), end(GraphMap[parent]) );
   RelocateVertex(vertex,0);
}

template <typename Tint>
void Graph<Tint>::EraseVertex(const Tint vertex){
   Tint parent = GraphMapRev[vertex];
   for(Tint i = 0; i< GraphMap[vertex].size(); i++){
     RelocateVertex(GraphMap[vertex][i],parent);
   }
   GraphMap[parent].erase( remove( begin(GraphMap[parent]), end(GraphMap[parent]), vertex ), end(GraphMap[parent]) );
}


template <typename Tint>
void Graph<Tint>::RelocateVertex(const Tint vertex, const Tint newvertexparent){
   GraphMapRev[vertex] = newvertexparent;
   GraphMap[newvertexparent].push_back(vertex);
}


template <typename Tint>
void Graph<Tint>::JoinVertex(const Tint vertexchild, const Tint vertexparent){
   AddVertex(vertexchild,vertexparent);
}


template <typename Tint>
void Graph<Tint>::DepthFirstTrav(const Tint vertex, Tint depth, Tint Flag){
  for(Tint i = 0;i< GraphMap[vertex].size();i++){ 
    if(Traversed.find(GraphMap[vertex][i]) != Traversed.end()){
      cerr << "Cycle detected! Edge:" << vertex << " -- " << GraphMap[vertex][i] << endl;
      continue;
    }
    if(Flag == 0)
      depth--;
    Flag = 0;
    Traversed[GraphMap[vertex][i]] = 1;
    Distance.push_back(++depth);
    TraversalPath.push_back(GraphMap[vertex][i]);
    if(GraphMap.find(GraphMap[vertex][i]) != GraphMap.end()){
      DepthFirstTrav(GraphMap[vertex][i], depth, 1);
     // Distance.push_back(depth);
     // TraversalPath.push_back(GraphMap[vertex][i]);
    }
  }
}


template <typename Tint>
void Graph<Tint>::Clean(){
  vector<Tint>().swap(Distance);
  vector<Tint>().swap(TraversalPath);
  Traversed.clear();
}

template <typename Tint>
vector<Tint>& Graph<Tint>::GetDepthVec(){
  return Distance;
}

template <typename Tint>
vector<Tint>& Graph<Tint>::GetVerticesVec(){
  return TraversalPath;
}


/* FM: Bad implementation */
template <typename Tint>
void Graph<Tint>::DmpGraph(vector<Tint>& parent, vector<Tint>& child){
  Tint i = 0;
  for ( auto it = GraphMap.begin(); it != GraphMap.end(); ++it ){ // extreamly inefficient 
    for(Tint j = 0; j< GraphMap[it->first].size();j++){
      if(parent.size() == i){
        parent.resize((parent.size()+1) * 2);
        child.resize((child.size()+1) * 2);
      }
      parent[i] = it->first;
      child[i++]  = GraphMap[it->first][j];
    }
  }
  parent.resize(i);
  child.resize(i);
}




src/include/Clique.hpp
/*
 * Clique.hpp
 * 
 * Copyright 2015 Robert Bakaric <rbakaric@irb.hr>
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 * 
 * 
 */

#include <iostream>
#include <vector>
#include <set>

using namespace std;


template <typename Tint>
class CliMax : protected Graph<Tint>{
   
/* Data structures */
   Tint MinClique;
   vector <vector <Tint>> MaxCliques;
   
/* Functions */
 
   /* Optimization: 
    * this would go much faster if vectors instead os sets were used 
    */
   void Expand(set<Tint>& Memb, set<Tint>& Subg, set<Tint>& Cand); 
   
   public: 

/* Constructor */
   CliMax(vector<Tint>& parent, vector <Tint>& child, Tint Min);
   
/* Destructor */
   ~CliMax();
   
/* Functions */

   void ComputeMaxCliques();
   vector <vector<Tint>>& GetMaxCliques();

/* Not finished: */   
   //ComputeMaxClique();
   
};




/* Constructors */
template <typename Tint>
CliMax<Tint>::CliMax(vector<Tint>& parent, vector <Tint>& child, Tint Min): Graph<Tint>(parent,child){
   MinClique = Min;   
}

/* Destructors */
template <typename Tint>
CliMax<Tint>::~CliMax(){
   vector <vector <Tint>>().swap( MaxCliques );
}

/* Functions */
template <typename Tint>
vector <vector<Tint>>& CliMax<Tint>::GetMaxCliques(){
   
   return MaxCliques;
   
}

/*  Implementation od CLIQUES by  Tomita et al. (2006) */
template <typename Tint>
void CliMax<Tint>::ComputeMaxCliques(){
   
   set<Tint> Memb;   
   set<Tint> Subg;
   set<Tint> Cand;

   for ( auto it = this->GraphMap.begin(); it != this->GraphMap.end(); ++it ){ 
      if (it->second.size() < MinClique-1) 
         continue;
      Subg.insert(it->first);
      Cand.insert(it->first);
         
   }
   Expand(Memb, Subg, Cand); 
}

template <typename Tint>
void CliMax<Tint>::Expand(set<Tint>&Memb, set<Tint>&Subg, set<Tint>&Cand){
   
   if (!Subg.size()){
     if (Memb.size() >= MinClique){
        vector<Tint> vec(Memb.begin(), Memb.end());
        MaxCliques.push_back(vec);
      }
     return;
   }
   
    Tint u = -1;
    Tint max = 0;
   
    /* a vertex 'u' that maximizes intersection */
    
     for(auto it = Subg.begin(); it != Subg.end(); it++){  // this needs further optimization 
        Tint intersec = 0;
        vector<Tint> & Children = this->GraphMap.find(*it)->second;
        if (Children.size() <= max) 
           continue;
        for(Tint i = 0; i< Children.size(); i++){
           if (Cand.find(Children[i]) != Cand.end()) 
              intersec++;
        }
        if (u == -1 || intersec > max){
            max = intersec;
            u = *it;
        }   	 
     }
     vector<Tint> & Children =  this->GraphMap.find(u)->second;
     set<Tint> tmp;
     
     for(Tint i = 0; i< Children.size(); i++) // I could do better here
       tmp.insert(Children[i]);
     
    /* compute EXT_u */
    set<Tint> Ext_u;
    for (auto it = Cand.begin(); it != Cand.end(); it++){
       if (tmp.find(*it) != tmp.end()) 
          continue;
       Ext_u.insert(*it);
    }
        
    while(Ext_u.size()>0){
       auto it = Ext_u.begin();
       vector<Tint> & Children  = this->GraphMap.find(*it)->second;
       set<Tint> vert;  
       for(Tint i = 0; i< Children.size(); i++)
          vert.insert(Children[i]);
          
       Memb.insert(*it);
       set<Tint> Subg_new;
       set<Tint> Cand_new;
       
       set_intersection(Subg.begin(), Subg.end(), vert.begin(), vert.end(), insert_iterator<set<Tint> >(Subg_new, Subg_new.begin()));
       set_intersection(Cand.begin(), Cand.end(), vert.begin(), vert.end(), insert_iterator<set<Tint> >(Cand_new, Cand_new.begin()));
       
       Expand(Memb, Subg_new, Cand_new);
       Memb.erase(*it);
       Cand.erase(*it);
       Ext_u.erase(it);
   }
}


/*  Left for implementation:
 * 
 *     Maximum Clique 
 * 
 *     + Though this can be derived from maximal ones, the algorithm is 
 *       quite inefficient iff the maximum one is required.
 * */



src/apps/CliMax.cpp
/*
 * CliMax.cpp
 * 
 * Copyright 2015 Robert Bakaric <rbakaric@irb.hr>
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 * 
 * 
 */


#include <iostream>
#include <fstream>
#include <vector>
#include <boost/program_options.hpp>

#include <Graph.hpp>
#include <Clique.hpp>

namespace po = boost::program_options;
using namespace std;



void PrintAcknowledgements(){

cout <<"***************************************************************************************"<<"\n";
cout <<"                                        CliMax                                         "<<"\n";
cout <<"                                          by                                           "<<"\n";
cout <<"                                    Robert Bakaric                                     "<<"\n\n";
cout <<"CONTACT:                                                                               "<<"\n";
cout <<" Code written and maintained by Robert Bakaric,                                        "<<"\n";
cout <<" email: rbakaric@irb.hr , bakaric@evolbio.mpg.de                                       "<<"\n\n";
cout <<"ACKNOWLEDGEMENT:                                                                       "<<"\n";
cout <<"     Tomita, E.,Tanaka, A. and  Takahashi, H. (2006) The worst-case time complexity    "<<"\n";
cout <<"     for generating all maximal cliques and computational experiments. Theor. Comput.  "<<"\n";
cout <<"     Sci. 363:28-42                                                                    "<<"\n";
cout <<"                                                                                       "<<"\n";
cout <<"LICENSE:                                                                               "<<"\n";
cout <<" The program is distributed under the GNU General Public License. You should have      "<<"\n";
cout <<" received a copy of the licence together  with this software. If not, see              "<<"\n";
cout <<" http://www.gnu.org/licenses/                                                          "<<"\n";
cout <<"***************************************************************************************"<<"\n\n\n";
}


template <typename INT, typename CHARA>
po::variables_map SetOptions(INT& argc, CHARA& argv){

    try {
        string version = "1.0";
        po::options_description desc("Allowed options");
        desc.add_options()
            ("help,h", "produce help message")
            ("version,v", "print version information")
            ("min_clique_size,c", po::value< int >()->default_value(2), "Minimum clique size")
            ("graph-file,i", po::value< string >(), "Adjecency list (tab seperated -> child <tab> parent)")
        ;

        po::positional_options_description p;
        p.add("input-file,i", -1);
        
        po::variables_map vm;
        po::store(po::command_line_parser(argc, argv).
                  options(desc).positional(p).run(), vm);
        po::notify(vm);
    
        if (vm.count("help")) {
            cout << "Usage: ./program [options]\n\n";
            PrintAcknowledgements();
            cout << desc;
            exit(0);
        }else if (vm.count("version")) {
            cout << "Program version:  " << version << "\n";
            exit(0);
        }
        if (!vm.count("graph-file")){
            cout << "Adjecensy list not provided \n";
            exit(0);
        }
        return vm;    
    }
    catch(std::exception& e)
    {
        cout << e.what() << "\n";
        exit(0);
    }    
}


template <typename Tint>
void ReadGraph(const string & file, vector<Tint>& parent, vector<Tint>& child){
  
  fstream fs;
  fs.open (file.c_str(), ios::in);
  if ( !fs.is_open())
    throw runtime_error ("Cannot open file: " + file );


  int i = 0; 
  while( !fs.eof()) {
    if(child.size() == i){
       child.resize((child.size()+1)*2);
       parent.resize((parent.size()+1)*2);
    }
    fs >> child[i];
    fs >> parent[i++] ;
  }  
 
  parent.resize(i-1);
  child.resize(i-1);
  fs.close();
}


int main (int argc, char** argv){
   
   
/* Get cmd Options */  
   po::variables_map arg;
   arg = SetOptions(argc,argv);

   string file =  arg["graph-file"].as<string>();
   int cutoff =  arg["min_clique_size"].as<int>();

   vector<int> parent;
   vector<int> child;
   
   try{
      
/* Load Graph */  
      ReadGraph(file, parent, child);

/* Make CliMax object */
      CliMax<int> clique(parent, child, cutoff);
  
/* Compute Maximal Cliques */  
      clique.ComputeMaxCliques();
      
/* Print cliques */
      vector<vector<int>> mc = clique.GetMaxCliques();
      cout << "MaxCliqSize\t" << "(Vertices)\n";
      for(int i = 0; i< mc.size(); i++){
         cout << mc[i].size() << "\t(" << mc[i][0];
         for(int j = 1; j< mc[i].size(); j++)
            cout << "," << mc[i][j] ;
         cout << ")\n";
      }
    
   }catch(runtime_error& e){
      cerr << e.what() << "\n";
   }
   
   return 0;
}
